src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs

```diff
@@ -130,95 +130,123 @@ public static SafeFreeCredentials AcquireCredentialsHandle(X509Certificate certi
 
-            int errorCode = SSPIWrapper.EncryptMessage(GlobalSSPI.SSPISecureChannel, securityContext, securityBuffer, 0);
+                Interop.SspiCli.SecBuffer* emptySecBuffer = &unmanagedBuffer[3];
+                emptySecBuffer->BufferType = SecurityBufferType.SECBUFFER_EMPTY;
+                emptySecBuffer->cbBuffer = 0;
+                emptySecBuffer->pvBuffer = IntPtr.Zero;
 
-            if (errorCode != 0)
-            {
-                if (GlobalLog.IsEnabled)
+                int errorCode = GlobalSSPI.SSPISecureChannel.EncryptMessage(securityContext, sdcInOut, 0);
+
+                if (errorCode != 0)
                 {
-                    GlobalLog.Print("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt ERROR" + errorCode.ToString("x"));
+                    if (GlobalLog.IsEnabled)
+                    {
+                        GlobalLog.Print("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt ERROR" + errorCode.ToString("x"));
+                    }
+
+                    resultSize = 0;
+                    return SecurityStatusAdapterPal.GetSecurityStatusPalFromNativeInt(errorCode);
                 }
-                resultSize = 0;
-            }
-            else
-            {
-                // The full buffer may not be used.
-                resultSize = securityBuffer[0].size + securityBuffer[1].size + securityBuffer[2].size;
-            }
 
-            return SecurityStatusAdapterPal.GetSecurityStatusPalFromNativeInt(errorCode);
+                Debug.Assert(headerSecBuffer->cbBuffer >= 0 && dataSecBuffer->cbBuffer >= 0 && trailerSecBuffer->cbBuffer >= 0);
+                Debug.Assert(checked(headerSecBuffer->cbBuffer + dataSecBuffer->cbBuffer + trailerSecBuffer->cbBuffer) <= output.Length);
+
+                resultSize = checked(headerSecBuffer->cbBuffer + dataSecBuffer->cbBuffer + trailerSecBuffer->cbBuffer);
+                return new SecurityStatusPal(SecurityStatusPalErrorCode.OK);
+            }
         }
 
-        public static SecurityStatusPal DecryptMessage(SafeDeleteContext securityContext, byte[] buffer, ref int offset, ref int count)
+        public static unsafe SecurityStatusPal DecryptMessage(SafeDeleteContext securityContext, byte[] buffer, ref int offset, ref int count)
         {
-            // Decryption using SCHANNEL requires four buffers.
-            SecurityBuffer[] decspc = new SecurityBuffer[4];
-            decspc[0] = new SecurityBuffer(buffer, offset, count, SecurityBufferType.SECBUFFER_DATA);
-            decspc[1] = new SecurityBuffer(null, SecurityBufferType.SECBUFFER_EMPTY);
-            decspc[2] = new SecurityBuffer(null, SecurityBufferType.SECBUFFER_EMPTY);
-            decspc[3] = new SecurityBuffer(null, SecurityBufferType.SECBUFFER_EMPTY);
-
-            Interop.SECURITY_STATUS errorCode = (Interop.SECURITY_STATUS)SSPIWrapper.DecryptMessage(
```

Comment:
Yeah, from SslStream's initialization and then also from a bunch of code paths in NegotiateStream.  I understand the reason why these helpers were put in place, but in the fullness of time it'd probably be nice to either revamp them to be more efficient or to remove them entirely; as this PR shows, they add very non-trivial overhead.